test/helpers/macos/applescript.js

```diff
@@ -0,0 +1,126 @@
+'use strict';
+
+const assert = require('assert');
+
+const {
+  applescript,
+  execScript,
+  renderScript,
+  retryOnTimeout,
+  runScript,
+} = require('../../../lib/helpers/macos/applescript.js');
+const { parseCodePoints } = require('../../../lib/helpers/macos/parseCodePoints.js');
+const { keyCodeCommand } = require('../../../lib/helpers/macos/keyCodeCommand.js');
+
+suite('helpers/macos/applescript', () => {
+  testEach(
+    'parseCodePoints',
+    ({ description }) => description,
+    ({ keyCombination, command, throws }) => {
+      if (throws) {
+        assert.throws(() => parseCodePoints(keyCombination), throws);
+      } else {
+        assert.deepEqual(parseCodePoints(keyCombination), command);
+      }
+    },
+    [
+      {
+        description: 'space',
+        keyCombination: [' '],
+        command: keyCodeCommand(['space'], []),
+      },
+      {
+        description: 'option-a',
+        keyCombination: ['\ue00a', 'a'],
+        command: keyCodeCommand(['a'], ['option']),
+      },
+      {
+        description: 'command-option-shift-b',
+        keyCombination: ['\ue03d', '\ue00a', '\ue008', 'b'],
+        command: keyCodeCommand(['b'], ['command', 'option', 'shift']),
+      },
+      { description: 'throws on null', keyCombination: ['\ue000'], throws: /unknown key/ },
+      {
+        description: 'throws on invalid code',
+        keyCombination: ['df daf% ?'],
+        throws: /Invalid key code/,
+      },
+    ],
+  );
+
+  testEach(
+    'renderScript',
+    ({ description }) => description,
+    ({ command, script }) => {
+      assert.deepEqual(renderScript(command), applescript(script));
+    },
+    [
+      {
+        description: 'space',
+        command: keyCodeCommand(['space'], []),
+        script:
+          'with timeout of 5 seconds\n' +
+          '    tell application "System Events"\n' +
+          '        key code 49\n' +
+          '    end tell\n' +
+          'end timeout',
+      },
+      {
+        description: 'option-a',
+        command: keyCodeCommand(['a'], ['option']),
+        script:
+          'with timeout of 5 seconds\n' +
+          '    tell application "System Events"\n' +
+          '        key code 0 using option down\n' +
+          '    end tell\n' +
+          'end timeout',
+      },
+      {
+        description: 'command-option-shift-b',
+        command: keyCodeCommand(['b'], ['command', 'option', 'shift']),
+        script:
+          'with timeout of 5 seconds\n' +
+          '    tell application "System Events"\n' +
+          '        key code 11 using {command down, option down, shift down}\n' +
+          '    end tell\n' +
+          'end timeout',
+      },
+    ],
+  );
+
+  test('execScript', async () => {
+    assert.equal(await execScript(applescript('return "hello"')), 'hello\n');
+  });
+
+  test('retryOnTimeout', async () => {
```

Comment:
We should have a test for the branch that concerns an unexpected error from the "action" function and a test for a true timeout.

Reply:
By true timeout do you mean a timeout produced by `osascript` like through `execScript`? I'm not sure there is a way to reliably do this. I learned while trying to do this that Applescript's `with timeout` doesn't set a timeout for the duration of executing its body, but sets the timeout used by each statement inside the body from the default (of 120 seconds) to the new value.

For example I tried it with `delay 2` inside `with timeout of 1 seconds`. From what I can guess `delay` doesn't use the timeout (the default of the new one specified this way).

-----

I added a `await new Promise(resolve => setTimeout(resolve, 1)` to the original test.

Reply:
Apologies; my request was a little vague. The test you've added named 'rethrows timeout after too many retries' is the test I was looking for, so we're good.